### PR TITLE
Add calibration parameters for multiple batteries

### PR DESCRIPTION
**Pull request by ItsTimmy**
**Describe problem solved by the proposed pull request**
Support for multiple batteries was improved in #12034 , but there is still the issue that different batteries all use the same calibration parameters.

**Test data / coverage**
On Pixhawk 4, multiple batteries with different levels of charge and numbers of cells were tested.

This PR makes changes that affect all different boards, so more testing should be done on different hardware.

**Describe your preferred solution**
Improved the Battery library by adding the ability to instantiate different batteries with different sets of parameters. Split power reading in `sensors` module into a separate class for readability. 

**Additional context**
In the near future, support for digital power meters will be expanded, continuing the work done in #11755. This refactoring should make supporting this new source of battery data, as well as any future new sources, easier.

### CODE REVIEW

**Comment by ItsTimmy** [comment]
@feemi Follow-up from our discussion in #12034: This PR will hopefully address the issues you were having before. Would you mind testing it with your setup?

**Comment by hamishwillee** [comment]
@ItsTimmy This will have a docs impact and possibly quite a complex QGC UI change. Please ping me when you think it is time to engage re docs. The main doc to change would be : http://docs.px4.io/master/en/config/battery.html

And presumably we may need to change the failsafe behaviour too: http://docs.px4.io/master/en/config/safety.html#low-battery-failsafe

**Comment by dagar** [comment]
This is looking good, but I think we should briefly take a step back and try to sketch out a rough overall design that factors in all the current requirements.
 
 - different battery types (analog bricks, digital bricks, smart batteries)
        - battery_status message
 - multiple instances
 - overall system estimate (reporting, failsafe handling, etc)
 - system power architecture
     - if you have multiple batteries what's the overall percentage?
 - refactoring to cleanup the mess scattered across each platform adc and the sensors module
 
@ItsTimmy @davids5 let's schedule a call.

**Comment by dagar** [comment]
Related https://github.com/PX4/Firmware/issues/12033

**Comment by ItsTimmy** [comment]
Even though we don't have the multi-battery requirements worked out completely, I think this PR is still ready for review and merge. It does not affect the safety-critical decisions like low-battery failsafe behavior and reporting of overall system battery percentage. These are still open issues, but they can be addressed in a separate PR.

File: src/lib/battery/battery.h [hunk 1]
```diff
@@ -31,91 +31,157 @@
...
+#pragma once
+
+#include "battery_base.h"
+
+#include <px4_log.h>
+#include <math.h>
+
 /**
  * @file battery.h
+ * Basic implementation of BatteryBase. Batter1 is calibrated by BAT1_* parameters. Battery2 is calibrated
```
Comment (julianoes) [inline src/lib/battery/battery.h]: `Batter1` -> `Battery1`.

File: src/lib/battery/battery_base.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	 * @param current_raw Voltage of current sense resistor, in Amps
+	 * @param timestamp Time at which the ADC was read (use hrt_absolute_time())
+	 * @param selected_source This battery is on the brick that the selected source for selected_source
+	 * @param priority: The brick number -1. The term priority refers to the Vn connection on the LTC4417
+	 * @param throttle_normalized Throttle of the vehicle, between 0 and 1
+	 * @param armed Arming state of the vehicle
+	 * @param shouldPublish If True, this function published a battery_status uORB message.
+	 */
+	void updateBatteryStatus(hrt_abstime timestamp, float voltage_v, float current_a, bool connected,
+				 bool selected_source, int priority, float throttle_normalized, bool armed, bool shouldPublish = true);
```
Comment (julianoes) [inline src/lib/battery/battery_base.h]: Can we remove the default for `shouldPublish`? And call it `should_publish`. It's easy to get these bools wrong when there are many args, and if there are defaults it's even trickier.

File: src/lib/battery/battery_base.cpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+	ModuleParams(nullptr),
+	_warning(battery_status_s::BATTERY_WARNING_NONE),
+	_last_timestamp(0)
+{
+}
+
+void
+BatteryBase::reset()
+{
+	memset(&_battery_status, 0, sizeof(_battery_status));
```
Comment (julianoes) [inline src/lib/battery/battery_base.cpp]: If you set all values, you don't need the `memset`.

File: src/lib/battery/battery_base.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+BatteryBase::filterVoltage(float voltage_v)
+{
+	if (!_battery_initialized) {
+		_voltage_filtered_v = voltage_v;
+	}
+
+	// TODO: inspect that filter performance
+	const float filtered_next = _voltage_filtered_v * 0.99f + voltage_v * 0.01f;
+
+	if (PX4_ISFINITE(filtered_next)) {
```
Comment (julianoes) [inline src/lib/battery/battery_base.cpp]: Why would this ever not be finite?

File: src/lib/battery/battery_params_1.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2012-2017 PX4 Development Team. All rights reserved.
```
Comment (julianoes) [inline src/lib/battery/battery_params_1.c]: 2012-2019

File: src/lib/battery/battery_params_common.c [hunk 6]
```diff
@@ -1,6 +1,6 @@
 /****************************************************************************
  *
- *   Copyright (c) 2012-2017 PX4 Development Team. All rights reserved.
+ *   Copyright (c) 2013-2017 PX4 Development Team. All rights reserved.
```
Comment (julianoes) [inline src/lib/battery/battery_params_common.c]: 2012-2019 (I agree these are annoying).

File: src/modules/sensors/analog_battery.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+ * Whether the ADC channel for the voltage of this battery is valid.
+ * Corresponds to BOARD_BRICK_VALID_LIST
+ */
+bool AnalogBattery::is_valid()
+{
+#ifdef BOARD_BRICK_VALID_LIST
+	bool valid[BOARD_NUMBER_BRICKS] = BOARD_BRICK_VALID_LIST;
+	return valid[_get_brick_index()];
+#else
+	// TODO: Maybe return false instead?
```
Comment (julianoes) [inline src/modules/sensors/analog_battery.cpp]: What sort of setup would that be? :thinking:

File: src/modules/sensors/analog_battery.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+float
+AnalogBattery::_get_cnt_v_volt()
+{
+	float val = _get_cnt_v_volt_raw();
+
+
+	if (val < 0.0f) {
+		// TODO: This magic constant was hardcoded into sensors.cpp before I did the refactor. I don't know
+		//  what the best way is to make it not a magic number.
+		return 3.3f / 4096.0f;
```
Comment (julianoes) [inline src/modules/sensors/analog_battery.cpp]: What does the number mean?

File: src/modules/sensors/power.h [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+public:
+	Power();
+
+	/**
+	 * Updates the measurements of each battery.
+	 *
+	 * If the parameter `BAT_SOURCE` == 0, this function will also publish an instance of the uORB topic
+	 * `battery_status` for each battery. For reasons of backwards compability, instance 0 will always be the
+	 * primary battery. However, this may change in the future! In the future, please use orb_priority() to find
+	 * the primary battery.
```
Comment (julianoes) [inline src/modules/sensors/power.h]: :+1:

File: src/lib/battery/battery_base.cpp (hunk 3 above)
Comment (ItsTimmy) [reply src/lib/battery/battery_base.cpp]: There are a lot of fields that are not being set. And I am concerned that if the `battery_status` message ever changes (which it might have to, to support smart batteries), the new fields may go uninitialized. But if you'd like, I can still change it to set every field individually.

File: src/lib/battery/battery_base.cpp (hunk 4 above)
Comment (ItsTimmy) [reply src/lib/battery/battery_base.cpp]: If voltage_v is for some reason NaN, I guess. I did not write this part, it was in the original battery library, so I assumed there was a reason for it.

File: src/modules/sensors/analog_battery.cpp (hunk 7 above)
Comment (ItsTimmy) [reply src/modules/sensors/analog_battery.cpp]: I'm not sure. This is replicating the original mess of #if's in Sensors.cpp. I found that if I tried to simplify any of it, some builds would fail, so presumably the original author of that part of Sensors.cpp (I think @bkueng) had a good reason for this.

If you'd like, I can investigate this particular case, and find where it is necessary.

File: src/modules/sensors/analog_battery.cpp (hunk 8 above)
Comment (ItsTimmy) [reply src/modules/sensors/analog_battery.cpp]: The number of ADC counts per volt. Looking back at this right now, I guess I should at least extract it as a constant somewhere.

File: src/lib/battery/battery_base.h (hunk 2 above)
Comment (ItsTimmy) [reply src/lib/battery/battery_base.h]: I spent too much time last semester in Python, where any function can have a hundred optional parameters!

File: src/modules/sensors/analog_battery.cpp (hunk 8 above)
Comment (davids5) [reply src/modules/sensors/analog_battery.cpp]: Value like this need to come from the board config it is dependent on the ADC resolution and the reference voltage. You can use the override pattern that is in board_common.h and default it to  3.3V and 12 Bits.

**Comment by davids5** [comment]
@ItsTimmy @julianoes @dagar -  I want to ensure the number of bricks logic is well understood from a HW perspective across platforms.  I still feel we need to have that meeting @dagar has arranged.

File: src/lib/battery/battery_base.cpp (hunk 3 above)
Comment (julianoes) [reply src/lib/battery/battery_base.cpp]: Ok fair.

File: src/modules/sensors/analog_battery.cpp (hunk 7 above)
Comment (julianoes) [reply src/modules/sensors/analog_battery.cpp]: I see. Leave as is then.

**Comment by julianoes** [comment]
@dagar should we merge this or wait for 1.10?

File: src/lib/battery/battery.h [hunk 10]
```diff
@@ -31,91 +31,157 @@
...
+	 * @tparam P2 Type of the second parameter
+	 * @tparam T Data type for the default value
+	 * @param oldParam Reference to the old parameter, as a ParamFloat<...>, ParamInt<...>, or ParamBool<...>
+	 * @param newParam Reference to the new paramater, as a ParamFloat<...>, ParamInt<...>, or ParamBool<...>
+	 * @param name The name of the parameter, WITHOUT the "BAT_" or "BAT1_" prefix. This is used only for logging.
+	 * @param defaultValue Default value of the parameter, as specified in PARAM_DEFINE_*(...)
 	 */
-	void reset(battery_status_s *battery_status);
+	template<class P1, class P2, typename T> static void
+	migrateParam(P1 &oldParam, P2 &newParam, const char *name, T defaultValue)
```
Comment (dagar) [inline src/lib/battery/battery.h]: Let's think about making this a generic parameter migration mechanism system wide.
@julianoes @davids5

Comment (hamishwillee) [reply src/lib/battery/battery.h]: If this happens let me know - sounds like a feature to document.
Will feature need to migrate parameters by pattern? ie BATTERY_n_CONFIG

**Comment by ItsTimmy** [comment]
@dagar @julianoes I think this is still ready to merge. It is technically consistent with the changes in mavlink/mavlink#1182, but there is a lot of work still to be done concerning configuration. In this PR, the mapping of physical battery to battery "slot" (Discussed in that Mavlink PR) is based solely on which power connector the power module is connected to.

In the future, as more power modules are supported (Such as the INA226 in #12673), we will need to decide how we map physical batteries to battery slots. But as that is outside the scope of this particular PR, I propose we address it in another PR.

**Comment by julianoes** [comment]
@ItsTimmy can you please rebase this branch?

**Comment by ItsTimmy** [comment]
It was overflowing flash on `px4_fmu-v5_stackcheck`, so I removed some modules from the build. Can you please double check that for me @julianoes ?

**Comment by julianoes** [comment]
Right, so now it's the normal "doesn't fit on fmu-v2".

**Comment by julianoes** [comment]
@dagar what should we do about the flash size for this one?

**Comment by ItsTimmy** [comment]
This has now been rebased to resolve the conflicts with the new `battery_status` module.

File: src/lib/battery/battery_params_common.c [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * @file battery_params.c
+ *
+ * Parameters defined by the battery lib, shared between all batteries.
+ *
+ * @author Julian Oes <julian@oes.ch>
+ */
+
+#include <px4_config.h>
+#include <parameters/param.h>
```
Comment (dagar) [inline src/lib/battery/battery_params_common.c]: These aren't actual C files by the way.

Comment (dagar) [inline src/lib/battery/battery_params_common.c]:
```suggestion
```

File: src/modules/battery_status/battery_status.cpp [hunk 12]
```diff
@@ -138,7 +139,19 @@ class BatteryStatus : public ModuleBase<BatteryStatus>, public ModuleParams, pub
 	uORB::Subscription	_vcontrol_mode_sub{ORB_ID(vehicle_control_mode)};		/**< vehicle control mode subscription */
 
 	orb_advert_t	_battery_pub[BOARD_NUMBER_BRICKS] {};			/**< battery status */
-	Battery		_battery[BOARD_NUMBER_BRICKS];			/**< Helper lib to publish battery_status topic. */
+
+	AnalogBattery1 _battery0;
+
+#if BOARD_NUMBER_BRICKS > 1
+	AnalogBattery2 _battery1;
```
Comment (dagar) [inline src/modules/battery_status/battery_status.cpp]: We should be able to split this into different modules by type and not have these ifdefs.

File: src/drivers/driver_framework_wrapper/df_ltc2946_wrapper/df_ltc2946_wrapper.cpp [hunk 13]
```diff
@@ -70,9 +69,7 @@ class DfLtc2946Wrapper : public LTC2946
 private:
 	int _publish(const struct ltc2946_sensor_data &data);
 
-	orb_advert_t	_battery_pub{nullptr};
-	battery_status_s _battery_status{};
-	Battery _battery{};
+	Battery1 _battery{};
```
Comment (dagar) [inline src/drivers/driver_framework_wrapper/df_ltc2946_wrapper/df_ltc2946_wrapper.cpp]: It seems a bit weird to have the instance explicitly in the type like this. Would a generic Battery<> or AnalogBattery<> work?

**Comment by dagar** [comment]
Where does the overall system power architecture fit into this? If I have 2 batteries connected don't I need to know more overall to provide a sensible overall percentage estimate for the system?

File: src/modules/battery_status/battery_status.cpp (hunk 12 above)
Comment (ItsTimmy) [reply src/modules/battery_status/battery_status.cpp]: I don't understand, these are all analog batteries in this module. The only other type of battery I have worked on is with an I2C power monitor in #12673 , and it is in a completely separate module. The reason these `#if`s are needed here is because, if a board only supports 1 power brick, but I declare both `AnalogBattery`s, then the compiler will complain about the unused variable.

**Comment by ItsTimmy** [comment]
You are correct in that there are a lot of open questions about how to handle multiple batteries. If you look in mavlink/mavlink#1182, you can see the ongoing discussion.

This PR does not address this in any way. Because that discussion is not fully resolved (at the time I finished up this PR in July), I chose to keep the same end behavior, but start refactoring things to get ready for more comprehensive support of multiple batteries. To that end, this PR has two goals:
- Add support for multiple different calibration values for different batteries. For example, two analog power modules might have slightly different voltage dividers and will need to be independently calibrated
- Refactor the battery library in such a way that multiple instances of new types of batteries can be added more easily. In the future we will need to support smart batteries, external batteries coming in over MAVLink, and perhaps others. I have demonstrated this by adding support for I2C digital power modules, in #12673 .

To summarize: This PR does not change the behavior of the power system, other than by adding the option to independently configure different batteries. It was intended to make it easier to add more battery types in the future.

**Comment by dagar** [comment]
This might need a bit of thought, but it's probably worth moving the parameter migration to the parameter system itself. On import we could catch the old values (by string) and set them in the new parameter. 

https://github.com/PX4/Firmware/blob/b943bd72ab20ecab155976d74986926d1c0349ea/src/lib/parameters/parameters.cpp#L1183

File: src/modules/battery_status/battery_status.cpp [hunk 14]
```diff
@@ -264,8 +277,8 @@ BatteryStatus::adc_poll()
 	bool  valid_chan[BOARD_NUMBER_BRICKS] = BOARD_BRICK_VALID_LIST;
 
 	/* Per Brick readings with default unread channels at 0 */
-	float bat_current_a[BOARD_NUMBER_BRICKS] = {0.0f};
-	float bat_voltage_v[BOARD_NUMBER_BRICKS] = {0.0f};
+	int32_t bat_current_adc_readings[BOARD_NUMBER_BRICKS] = {0};
```
Comment (dagar) [inline src/modules/battery_status/battery_status.cpp]:
```suggestion
	int32_t bat_current_adc_readings[BOARD_NUMBER_BRICKS]{};
```

File: src/lib/battery/battery.h (hunk 10 above)
Comment (julianoes) [reply src/lib/battery/battery.h]: @dagar we can pull this out later, I'd say.

File: src/lib/battery/battery_params_1.c [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+/**
+ * @file battery_params_1.c
+ * @author Timothy Scott <timothy@auterion.com>
+ *
+ * Defines parameters for Battery 1. For backwards compatibility, the
+ * parameter names do not have a "1" in them.
+ */
+
+#include <px4_platform_common/px4_config.h>
+#include <parameters/param.h>
```
Comment (julianoes) [inline src/lib/battery/battery_params_1.c]: As @dagar already wrote, you can remove these.

File: src/modules/battery_status/power.cpp [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * @file power.cpp
+ *
+ * @author Timothy Scott <timothy@auterion.com>
+ */
+#include "power.h"
+
+Power::Power() {}
+
+void Power::update(px4_adc_msg_t buf_adc[PX4_MAX_ADC_CHANNELS], int nchannels, float throttle, bool armed)
```
Comment (julianoes) [inline src/modules/battery_status/power.cpp]: I think the first argument can be `const`, right?

File: src/modules/battery_status/power.h [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+	 * If the parameter `BAT_SOURCE` == 0, this function will also publish an instance of the uORB topic
+	 * `battery_status` for each battery. For reasons of backwards compability, instance 0 will always be the
+	 * primary battery. However, this may change in the future! In the future, please use orb_priority() to find
+	 * the primary battery.
+	 * @param buf_adc Buffer of ADC readings
+	 * @param nchannels Number of valid ADC readings in `buf_adc`
+	 * @param throttle Normalized throttle (between 0 and 1, or maybe between -1 and 1 in the future)
+	 * @param armed True if the vehicle is armed
+	 */
+	void update(px4_adc_msg_t buf_adc[PX4_MAX_ADC_CHANNELS], int nchannels, float throttle, bool armed);
```
Comment (julianoes) [inline src/modules/battery_status/power.h]: And of course `const` here too then.

File: src/modules/simulator/simulator.h [hunk 18]
```diff
@@ -236,10 +234,24 @@ class Simulator : public ModuleParams
...
 	hrt_abstime _last_sitl_timestamp{0};
+	hrt_abstime _last_battery_timestamp{0};
 
-	// Lib used to do the battery calculations.
-	Battery _battery {};
-	battery_status_s _battery_status{};
+	// Because the simulator doesn't actually care about the real values of these, just stick
+	// in some good defaults.
+	// This is an anonymous class, because BatteryBase is abstract and can't directly be instantiated.
+	class : public BatteryBase
```
Comment (julianoes) [inline src/modules/simulator/simulator.h]: I have never seen this before :astonished:. I guess it works?

File: src/modules/battery_status/power.h (hunk 17 above)
Comment (ItsTimmy) [reply src/modules/battery_status/power.h]: I ended up not using these files, because of the refactor. I copied them over while rebasing and forgot to delete them.

File: src/modules/simulator/simulator.h (hunk 18 above)
Comment (ItsTimmy) [reply src/modules/simulator/simulator.h]: Yep! It's the same way you can declare an anonymous struct type:
```
struct {
  int a;
  float b;
} anonymous_struct;
anonymous_struct.a = 1;
```

Comment (julianoes) [reply src/modules/simulator/simulator.h]: Ah!

File: src/modules/battery_status/power.h (hunk 17 above)
Comment (julianoes) [reply src/modules/battery_status/power.h]: aha.

**Comment by julianoes** [comment]
@dagar would be good to get this in! It's been a while :smile:.

**Comment by ItsTimmy** [comment]
@hamishwillee As far as I know, no, there is currently no way to do templated parameters, unfortunately.

And as for support in QGC, I am not sure if there is a specific roadmap. There is the open PR mavlink/qgroundcontrol#7575, which addresses displaying the battery readings, but not calibrating separate batteries.

**Comment by hamishwillee** [comment]
@ItsTimmy FYI, templated param definitions are possible https://github.com/PX4/Firmware/blob/master/src/modules/mavlink/module.yaml#L20 but not templated code. @bkueng is the expert.

**Comment by ItsTimmy** [comment]
@hamishwillee I was not aware of that, thanks for the info! I spoke with Beat, so I now know how to use that feature. Incorporating it into this PR will be a bit tricky though, since some parameters are defined outside of a module (in the battery library). I will look into that though, and consider adding it here.

**Comment by hamishwillee** [comment]
@ItsTimmy All good with me - I just wanted to make you aware of the option. I have kicked off https://github.com/PX4/Devguide/pull/898 to get the Yaml/multi instance stuff documented.

**Comment by ItsTimmy** [comment]
I have updated this with 3 important differences:
- Parameters are now instanced with YAML files
- Refactored to decrease code size
- Parameter migration has been improved: Old and new parameters are now kept in sync in both directions. If either parameter is changed, the other is immediately updated to reflect this. If the deprecated parameter was changed, then a warning is displayed.

I still have some cleanup and testing to do on this. There are still some relics of the old way this was written.

File: src/lib/battery/battery.cpp [hunk 19]
```diff
@@ -32,46 +32,96 @@
...
 
-Battery::Battery() :
-	ModuleParams(nullptr),
+Battery::Battery(int index) :
+	_index(index < 1 || index > 9 ? 1 : index),
 	_warning(battery_status_s::BATTERY_WARNING_NONE),
 	_last_timestamp(0)
 {
+	if (index > 9 || index < 1) {
+		PX4_ERR("Battery index must be between 1 and 9 (inclusive). Received %d. Defaulting to 1.", index);
```
Comment (julianoes) [inline src/lib/battery/battery.cpp]: I guess the alternative would just be an `assert`.

File: src/lib/battery/battery_params_deprecated.c [hunk 20]
```diff
@@ -142,7 +89,7 @@ PARAM_DEFINE_FLOAT(BAT_EMERGEN_THR, 0.05f);
 PARAM_DEFINE_FLOAT(BAT_V_LOAD_DROP, 0.3f);
 
 /**
- * Explicitly defines the per cell internal resistance
+ * This parameter is deprecated. Please use BAT1_R_INTERNAL instead.
```
Comment (julianoes) [inline src/lib/battery/battery_params_deprecated.c]: But `BAT1_R_INTERNAL` is not available yet, right?

Comment (ItsTimmy) [reply src/lib/battery/battery_params_deprecated.c]: https://github.com/ItsTimmy/Firmware/blob/612ae0c74341171d4e665dbbaaaacc97b46960d5/src/lib/battery/module.yaml#L61-L77

File: src/lib/battery/battery.h [hunk 21]
```diff
@@ -32,59 +32,179 @@
...
+		float v_load_drop_old;
+		float r_internal_old;
+		int source_old;
+	} _params;
+
+	battery_status_s _battery_status;
+
+	const int _index;
+
+	uORB::Subscription _param_sub{ORB_ID(parameter_update)};
```
Comment (bkueng) [inline src/lib/battery/battery.h]: Why do you handle this in Battery? I prefer the existing mechanism via `ModuleParams`, then you don't need it here as well.

Comment (ItsTimmy) [reply src/lib/battery/battery.h]: Because I am using instanced parameters in the yaml file, I was under the impression that I wouldn't be able to use the nice ModuleParams API. Is there some way to properly use the ModuleParams API here? I tried using it in my old solution, but it was very messy and required a different class for each instance, which increased code size.

This current solution is not ideal, but I am not aware of a better way to use instanced parameters.

File: src/lib/battery/CMakeLists.txt [hunk 22]
```diff
@@ -31,4 +31,6 @@
 #
 ############################################################################
 
-px4_add_library(battery battery.cpp)
+px4_add_library(battery battery.cpp battery.cpp)
+
+set_property(GLOBAL APPEND PROPERTY PX4_MODULE_CONFIG_FILES ${CMAKE_CURRENT_SOURCE_DIR}/module.yaml)
```
[the remark this answers is not in the document]
Comment (ItsTimmy) [reply src/lib/battery/CMakeLists.txt]: I don't think so.

`px4_add_library`: https://github.com/PX4/Firmware/blob/133a6e38cc788f854e4f6596a6da026d8b1f2a5f/cmake/px4_add_library.cmake#L40-L55

The relevant part of `px4_add_module`, which is not present in `px4_add_library`:
https://github.com/PX4/Firmware/blob/133a6e38cc788f854e4f6596a6da026d8b1f2a5f/cmake/px4_add_library.cmake#L40-L55

I could of course try to add this functionality to `px4_add_library`, but I did not think that was within the scope of this PR.

File: src/lib/battery/battery.h (hunk 21 above)
Comment (bkueng) [reply src/lib/battery/battery.h]: You can still use it for the notification, just override `updateParams`. The rest is better done via C API at the moment, you are right.

Comment (ItsTimmy) [reply src/lib/battery/battery.h]: Oh OK, now I understand!

File: src/lib/battery/CMakeLists.txt (hunk 22 above)
[the remark this answers is not in the document]
Comment (bkueng) [reply src/lib/battery/CMakeLists.txt]: I'm aware the functionality is missing ATM, which is what I am asking for to implement. If you don't want to do that here, that's ok, but please add a TODO as a comment.

[the remark this answers is not in the document]
Comment (ItsTimmy) [reply src/lib/battery/CMakeLists.txt]: I made an attempt at this in #13633, but it was not easy. I am not great with CMake, so there may have been a simpler solution that I was not aware of.

File: src/modules/simulator/simulator.h [hunk 23]
```diff
@@ -236,10 +235,25 @@ class Simulator : public ModuleParams
...
+	hrt_abstime _last_battery_timestamp{0};
 
-	// Lib used to do the battery calculations.
-	Battery _battery {};
-	battery_status_s _battery_status{};
+	class : public Battery
+	{
+		virtual bool updateParams(bool force) override
+		{
+			_params.v_empty = 3.5;
```
Comment (dagar) [inline src/modules/simulator/simulator.h]: Always use f for float literals.

Comment (dagar) [inline src/modules/simulator/simulator.h]:
```suggestion
			_params.v_empty = 3.5f;
```

File: src/modules/battery_status/analog_battery.cpp [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+
+		param_get(_analog_param_handles.cnt_v_volt, &_analog_params.cnt_v_volt);
+		param_get(_analog_param_handles.cnt_v_curr, &_analog_params.cnt_v_curr);
+		param_get(_analog_param_handles.v_offs_cur, &_analog_params.v_offs_cur);
+
+
+		if (_analog_params.cnt_v_volt < 0.0f) {
+			// TODO: This magic constant was hardcoded into sensors.cpp before I did the refactor. I don't know
+			//  what the best way is to make it not a magic number.
+			_analog_params.cnt_v_volt = DEFAULT_VOLTS_PER_COUNT;
```
Comment (dagar) [inline src/modules/battery_status/analog_battery.cpp]: It's slightly different in current master.

https://github.com/PX4/Firmware/blob/bd98a4eebbee750a4db007b03bcd1e40896cb5d5/src/modules/battery_status/parameters.cpp#L71-L73

**Comment by ItsTimmy** [comment]
@dagar @julianoes @bkueng This is ready for review.

**Comment by ItsTimmy** [comment]
I did some more testing on this latest version. The following tests all worked as intended:

 - Plugged in 2 batteries (with different cell counts) to Pixhawk 4 with analog power module
 - Calibrated battery 1 with QGC
 - Calibrated battery 2 manually (using `param set`)
 - Set values of all `BAT_*` parameters and `BAT1_*` parameters to check that migration worked correctly

File: src/lib/battery/battery.h [hunk 25]
```diff
@@ -37,54 +37,200 @@
...
+		T previous_old_val = *old_val;
+		T previous_new_val = *new_val;
+
+		param_get(old_param, old_val);
+		param_get(new_param, new_val);
+
+		if (!firstcall) {
+			if ((float) fabs((float) *old_val - (float) previous_old_val) > FLT_EPSILON
+			    && (float) fabs((float) *old_val - (float) *new_val) > FLT_EPSILON) {
+				// TODO fix this error message. I was getting hardfaults while using this message, but they stopped
```
Comment (bkueng) [inline src/lib/battery/battery.h]: You can remove it completely, as a user would not see it.
The hardfault should either be a stack overflow or null-pointer dereference (`param_name`).

File: src/modules/battery_status/analog_battery.cpp [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+	bool connected = voltage_v > BOARD_ADC_OPEN_CIRCUIT_V &&
+			 (BOARD_ADC_OPEN_CIRCUIT_V <= BOARD_VALID_UV || is_valid());
+
+
+	Battery::updateBatteryStatus(timestamp, voltage_v, current_a, connected,
+				     selected_source, priority, throttle_normalized, armed, _params.source == 0);
+}
+
+/**
+ * Whether the ADC channel for the voltage of this battery is valid.
```
Comment (bkueng) [inline src/modules/battery_status/analog_battery.cpp]: The docs in the header is enough.

File: src/modules/simulator/simulator.h [hunk 27]
```diff
@@ -237,10 +236,24 @@ class Simulator : public ModuleParams
...
+			_params.v_empty = 3.5f;
+			_params.v_charged = 4.05f;
+			_params.n_cells = 4;
+			_params.capacity = 10.0f;
+			_params.v_load_drop = 0.0f;
+			_params.r_internal = 0.0f;
+			_params.low_thr = 0.15f;
+			_params.crit_thr = 0.07f;
+			_params.emergen_thr = 0.05f;
+			_params.source = 0;
```
Comment (bkueng) [inline src/modules/simulator/simulator.h]: I'd still call `Battery::updateParams()` here.

File: src/lib/battery/battery.h [hunk 28]
```diff
@@ -37,54 +37,200 @@
...
+ * BatteryBase is a base class for any type of battery.
+ *
+ * You can use this class on its own. Or, if you need to implement a custom battery type,
+ * you can inherit from this class. See, for example, src/modules/battery_status/AnalogBattery.h
+ */
 class Battery : public ModuleParams
 {
 public:
-	Battery();
+	Battery(int index = 1, ModuleParams *parent = nullptr);
```
Comment (bkueng) [inline src/lib/battery/battery.h]: I would not default to NULL for the `parent`, to ensure it's set.

File: src/modules/simulator/simulator.h (hunk 27 above)
Comment (ItsTimmy) [reply src/modules/simulator/simulator.h]: Why? In the original version of this, the parameters were not used. They were just hard-coded values like this. If I properly use the parameters, won't that change the behavior of the simulator?

Comment (bkueng) [reply src/modules/simulator/simulator.h]: Not for now, but in case parameters are used/added in future to the base class.
